src/field/generated_fieldops.cxx

```diff
@@ -0,0 +1,1668 @@
+
+  for (int i = 0; i < len; ++i) {
+    result[i] = lhs[i] - rhs[i];
+  }
+}
+
+// Provide the C++ wrapper for subtraction of Field3D and Field3D
+Field3D operator-(const Field3D &lhs, const Field3D &rhs) {
+  Indices i{0, 0, 0};
+  Mesh *localmesh = lhs.getMesh();
+
+  ASSERT1(localmesh == rhs.getMesh());
+
+  Field3D result(localmesh);
+  result.allocate();
+  checkData(lhs);
+  checkData(rhs);
+
+  autogen_Field3D_Field3D_Field3D_subtraction(&result[i], &lhs[i], &rhs[i],
+                                              localmesh->LocalNx * localmesh->LocalNy *
+                                                  localmesh->LocalNz);
+
+#if CHECK > 0
+  if (lhs.getLocation() != rhs.getLocation()) {
+    throw BoutException("Error in subtraction: fields at different locations. lhs is at "
+                        "%s, rhs is at %s!",
+                        strLocation(lhs.getLocation()), strLocation(rhs.getLocation()));
+  }
+#endif
+
+  result.setLocation(rhs.getLocation());
+
+  checkData(result);
+  return result;
+}
+
+// Provide the C function to update Field3D by subtraction with Field3D
+void autogen_Field3D_Field3D_subtraction(BoutReal *__restrict__ lhs,
+                                         const BoutReal *__restrict__ rhs, int len) {
+
+  for (int i = 0; i < len; ++i) {
+    lhs[i] -= rhs[i];
+  }
+}
+
+// Provide the C++ operator to update Field3D by subtraction with Field3D
+Field3D &Field3D::operator-=(const Field3D &rhs) {
+  // only if data is unique we update the field
+  // otherwise just call the non-inplace version
+  if (data.unique()) {
+    Indices i{0, 0, 0};
+
+    ASSERT1(fieldmesh == rhs.getMesh());
+
+    checkData(*this);
+    checkData(rhs);
+
+    autogen_Field3D_Field3D_subtraction(&(*this)[i], &rhs[i],
+                                        fieldmesh->LocalNx * fieldmesh->LocalNy *
+                                            fieldmesh->LocalNz);
+
+#if CHECK > 0
+    if (this->getLocation() != rhs.getLocation()) {
+      throw BoutException("Error in subtraction: fields at different locations. lhs is "
+                          "at %s, rhs is at %s!",
+                          strLocation(this->getLocation()),
+                          strLocation(rhs.getLocation()));
+    }
+#endif
+    checkData(*this);
+
+  } else {
+    (*this) = (*this) - rhs;
+  }
+  return *this;
+}
+
+// Do the actual multiplication of Field3D and Field2D
+void autogen_Field3D_Field3D_Field2D_multiplication(BoutReal *__restrict__ result,
+                                                    const BoutReal *__restrict__ lhs,
+                                                    const BoutReal *__restrict__ rhs,
+                                                    int nx, int ny, int nz) {
+
+  for (int x = 0; x < nx; ++x)
+    for (int y = 0; y < ny; ++y)
+      for (int z = 0; z < nz; ++z) {
+        result[z + nz * (y + ny * x)] = lhs[z + nz * (y + ny * x)] * rhs[y + x * ny];
```

Comment:
Performance wise has the use of a counter been tested (e.g. `uint i=0` outside the loops and then `result[i] = lhs[i]*rhs[i2d]; i++;` where `i2d` can either be calculated outside the innermost loop or could be a second counter that is incremented one level up from `i`)? This may or may not help.